File: app/src/main/java/com/kaiserpudding/howtheywrite/lesson/LessonListAdapter.java
package com.kaiserpudding.howtheywrite.lesson;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;
import com.kaiserpudding.howtheywrite.R;
import com.kaiserpudding.howtheywrite.model.Lesson;
import java.util.List;

public class LessonListAdapter extends RecyclerView.Adapter<LessonListAdapter.LessonViewHolder> {

  class LessonViewHolder extends RecyclerView.ViewHolder {

    private final TextView lessonItemView;

    private LessonViewHolder (Context context, View itemView) {
      super(itemView);
      itemView.setOnClickListener((v) -> {
        Toast.makeText(context, String.valueOf(getAdapterPosition()), Toast.LENGTH_SHORT).show();
          }
      );
      lessonItemView = itemView.findViewById(R.id.textView);
    }
  }

  private final LayoutInflater inflater;
  private final Context context;

  private List<Lesson> lessons;

  LessonListAdapter(Context context) {
    this.context = context;
    inflater = LayoutInflater.from(context);
  }

  @NonNull
  @Override
  public LessonViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
    View itemView = inflater.inflate(R.layout.lesson_recyclerview_item, parent, false);
    return new LessonViewHolder(context, itemView);
  }

  @Override
  public void onBindViewHolder(@NonNull LessonViewHolder holder, int position) {
    if (lessons != null) {
      Lesson current = lessons.get(position);
      //TODO proper string representation of lesson
      holder.lessonItemView.setText(current.toString());
    } else {
      holder.lessonItemView.setText(R.string.no_lesson);
    }
  }

  void setLessons(List<Lesson> lessons) {
    this.lessons = lessons;
    notifyDataSetChanged();
  }

  @Override
  public int getItemCount() {
    if (lessons != null)
      return  lessons.size();
      else return 0;
  }

}

File: HowTheyWrite_android/app/src/main/java/com/kaiserpudding/howtheywrite/lessonList/LessonListAdapter.kt
package com.kaiserpudding.howtheywrite.lessonList

import android.content.Context
import android.view.LayoutInflater
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import com.kaiserpudding.howtheywrite.R
import com.kaiserpudding.howtheywrite.model.Lesson
import com.kaiserpudding.howtheywrite.shared.multiSelect.MultiSelectAdapter

/**
 * Adapter for the recyclerView for lessions
 *
 * Classes using this must implement [MultiSelectAdapter.MultiSelectAdapterItemInteractionListener]
 * to handle interaction events with an item in the list.
 *
 * @property context
 * @property listener
 */
class LessonListAdapter(
        private val context: Context,
        listener: MultiSelectAdapterItemInteractionListener)
    : MultiSelectAdapter<Lesson>(listener), Filterable {

    private val inflater: LayoutInflater = LayoutInflater.from(context)
    override val viewHolderId = R.id.lessonTextView
    private val all = Lesson("All")
    private var filteredList: List<Lesson>?

    init {
        list = listOf(all)
        filteredList = list
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MultiSelectViewHolder {
        val itemView = inflater.inflate(R.layout.recyclerview_item_lesson, parent, false)
        return createViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: MultiSelectViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)

        val current = filteredList!![position]
        //TODO proper string representation of lesson
        holder.textView.text = current.toString()
    }

    internal fun setLessons(lessons: List<Lesson>) {
        val combined = mutableListOf(all)
        combined.addAll(lessons)
        list = combined
        filteredList = list
        notifyDataSetChanged()
    }

    internal fun resetLessons() {
        filteredList = list
        notifyDataSetChanged()
    }

    override fun getItemCount(): Int {
        return if (filteredList != null) filteredList!!.size
        else 0
    }

    override fun getMyItemId(position: Int): Long {
        return if (filteredList != null) filteredList!![position].id
        else 0
    }

    override fun getFilter(): Filter {
        return object : Filter() {
            override fun performFiltering(constraint: CharSequence): FilterResults {
                val results = FilterResults()
                if (constraint.isEmpty()) {
                    results.values = list
                } else {
                    val resultList = mutableListOf<Lesson>()
                    for (lesson in list!!) {
                        if (lesson.name.contains(constraint, true)) {
                            resultList.add(lesson)
                        }
                    }
                    results.values = resultList
                }
                return results
            }

            @Suppress("UNCHECKED_CAST")
            override fun publishResults(constraint: CharSequence, results: FilterResults) {
                filteredList = results.values as List<Lesson>?
                notifyDataSetChanged()
            }
        }
    }
}
